File: adapter-pattern/shapes/shape-adapter.ts
// Third party lib
class SquareDrawer {
  drawSquare(size: number): void {
    console.log(`Drawing a square with size of ${size}x${size}`);
  }
}

// Target interface (for end user consumption, client is expecting this)
interface RectangleDrawer {
  drawRectangle(width: number, height: number): void;
}

// Adapter that adapts SquareDrawer to the RectangleDrawer interface
class SquareToRectAdapter implements RectangleDrawer {
  private drawer: SquareDrawer;

  constructor(drawer: SquareDrawer) {
    this.drawer = drawer;
  }

  drawRectangle(width: number, height: number): void {
    if (width === height) {
      this.drawer.drawSquare(width);
    } else {
      console.log("Cannot draw rectangle with a square drawer :( \n");
    }
  }
}

// Usage
const squareDrawer = new SquareDrawer();
const adapter = new SquareToRectAdapter(squareDrawer);

adapter.drawRectangle(6, 6); // Works
adapter.drawRectangle(10, 20); // Fails

File: adapter-pattern/shapes/shape-adapter.rs
// Third-party lib
struct SquareDrawer;
impl SquareDrawer {
    fn draw_square(&self, size: u32) {
        println!("Drawing a square of size {}x{}", size, size);
    }
}

// Trait (interface) that the client is expecting
trait RectangleDrawer {
    fn draw_rectangle(&self, width: u32, height: u32);
}

// Adapter
struct SquareToRectangleAdapter<'a> {
    square_drawer: &'a SquareDrawer,
}

impl<'a> RectangleDrawer for SquareToRectangleAdapter<'a> {
    fn draw_rectangle(&self, width: u32, height: u32) {
        if width == height {
            self.square_drawer.draw_square(width);
        } else {
            println!("Cannot draw a rectangle with a square drawer!");
        }
    }
}

// Usage
fn main() {
    let square_drawer = SquareDrawer;
    let adapter = SquareToRectangleAdapter {
        square_drawer: &square_drawer,
    };

    adapter.draw_rectangle(6, 6); // Works
    adapter.draw_rectangle(10, 20); // Fails gracefully
}
